### DISCUSSION TITLE
Weight and Mass

### DISCUSSION THREAD

**Discussion by jonsberndt** [JSBSim-Team/jsbsim#368] [Ideas]
Not sure if this has been talked about before but I believe that the current code only permits a vehicle weight being entered via the aircraft config file:

```c++
  SetAircraftBaseInertias(ReadInertiaMatrix(document));
  if (document->FindElement("emptywt")) {
    EmptyWeight = document->FindElementValueAsNumberConvertTo("emptywt", "LBS");
  }
```
I am currently modeling a vehicle that will fly on another planet. That has involved changing the planet properties and will involve implementing a new atmosphere. I would like to add the ability to optionally enter a vehicle **mass** instead of weight. This would be an either/or proposition, of course. The Load() code might look something like this:

```c++
  SetAircraftBaseInertias(ReadInertiaMatrix(document));
  if (document->FindElement("emptywt")) {
    EmptyWeight = document->FindElementValueAsNumberConvertTo("emptywt", "LBS");
  } else if (document->FindElement("emptymass")) {
    EmptyWeight = local_gravity*document->FindElementValueAsNumberConvertTo("emptymass", "SLUGS");
  }
```
... or something like that.

Comments?

**Comment by bcoconni** [comment]
The name is misleading but the empty weight is actually the empty mass. The units of `<emptywt>` are pounds (of mass) and not pounds (of force). So unless I misunderstood your proposal, the feature you are looking for already exists.

In the code, the mass `Mass` is just a unit conversion of `Weight`:
https://github.com/JSBSim-Team/jsbsim/blob/868fcd59f4fea8bc051bd44b18a913c1d86de40c/src/models/FGMassBalance.cpp#L191-L194
However I agree that the name "empty weight" as it is currently used in JSBSim is improper since a weight is a force while the mass is, well, a mass :smile: The same is true about the general usage of the word "weight" in JSBSim which is improperly used...

**Reply by jonsberndt** [reply]
I don't think this is true. And the code (most or all of which I am the guilty party for ...) is very misleading. 

The variable Weight is a true weight (force), since the Mass variable that follows converts the Weight in JSBSim native pounds to the mass unit slugs. Look at the weight defined for the C172x aircraft in the c172x.xml file:
```xml
        <emptywt unit="LBS"> 1454.0 </emptywt>
```
This is the empty weight for the C172x in pounds force. It is subsequently converted to slugs in JSBSim when read in.

One can specify the empty weight for a vehicle in kilograms - which is an abomination 😄, since kg is a unit of mass. If an empty weight for a vehicle being modeled is specified in kg, it is assumed to be on the Earth surface and converted to pounds force. 

I think, though, that merely specifying the mass of the vehicle in slugs or kg does not solve the whole problem. I have added code in a custom JSBSim codebase that models the relevant planetary gravitation, but there are constants in JSBSim that need to be changed to make the broader mass properties code work.

There's probably some work that could be done for JSBSim to make it work for different planets (Moon, Mars, Titan, etc.)

**Reply by seanmcleod** [reply]
Had to refresh my non-SI knowledge :wink:

So there are 2 units of mass, lb (mass) and slug.

Looking at the units in the xml file, i.e. `LBS` I also assumed they referred to lb (mass) as opposed to lbf.

But I guess the following applies to JSBSim, from https://en.wikipedia.org/wiki/Pound_(force)

> In some contexts, the term "pound" is used almost exclusively to refer to the unit of force and not the unit of mass. In those applications, the preferred unit of mass is the slug, i.e. lbf⋅s2/ft. In other contexts, the unit "pound" refers to a unit of mass. The international standard symbol for the pound as a unit of mass is lb.

And as you mentioned there is the potential confusion when quoting kg for weight, e.g. for the [737-800](http://www.flugzeuginfo.net/acdata_php/acdata_7378_en.php#:~:text=Airliner%20%20%20Empty%20Weight%20%20%2041.413,%2065.317%20kg%20%20%20144.000%20lbs%20) which I looked up to compare with the JSBSim 737 model.

![image](https://user-images.githubusercontent.com/824856/103277767-6c11d580-49d2-11eb-9f9b-68752c6869d9.png)

So a mixture of lbf as a weight and mass in kg quoted as the equivalent weight when it's really the mass and not using `lbf` to be clear either!

**Reply by bcoconni** [reply]
> This is the empty weight for the C172x in pounds force.

How can you tell the difference since the number is the same between mass in pounds and weight in pounds-force ?
> One can specify the empty weight for a vehicle in kilograms - which is an abomination :smile:, since kg is a unit of mass. If an empty weight for a vehicle being modeled is specified in kg, it is assumed to be on the Earth surface and converted to pounds force.

Hmm, I disagree with that statement. Since the final conversion is in slugs (mass unit), then converting from kg to pounds (mass) then from pounds (mass) to slugs would give the exact same figure than we are currently getting so I don't think that would be an error.

It should also be noted that in European Type Certificates "weights" such as MTW, MTOW, MLW and MZFW are called "masses" (which avoid ambiguities between lbs and lbf) and are specified in kg. See for instance the [Boeing 737-900 Type Certificate issued by the EASA](https://www.easa.europa.eu/sites/default/files/dfu/EASA%20TCDS%20IM%20A%20120%20-%20rev%2020.pdf):
![B737-900_EASA_certified_masses](https://user-images.githubusercontent.com/7502904/104056091-3057e680-51f0-11eb-888f-f8afa7fd1702.png)
Fortunately, the figures expressed in lbs in the European Type Certificate match the values from the [Boeing 737-900 Type Certificate issued by the FAA](https://rgl.faa.gov/Regulatory_and_Guidance_Library/rgMakeModel.nsf/0/1c2b973615d72f9c8625861c00508676/$FILE/A16WE_Rev66.pdf) :smile::
![B737-900_FAA_certified_masses](https://user-images.githubusercontent.com/7502904/104055926-e2db7980-51ef-11eb-9b7c-7d8311e230a8.png)

**Comment by jonsberndt** [comment]
To further clarify, when I wrote the initial code I had no thought that it would ever be used to model flight on another planet. So, Earth properties were assumed. Since the weight of something depends on the acceleration of gravity, this code totally breaks down on a planet other than Earth:

```c++
  SetAircraftBaseInertias(ReadInertiaMatrix(document));
  if (document->FindElement("emptywt")) {
    EmptyWeight = document->FindElementValueAsNumberConvertTo("emptywt", "LBS");
  }

  Element *element = document->FindElement("location");
  while (element) {
    element_name = element->GetAttributeValue("name");
    if (element_name == "CG") vbaseXYZcg = element->FindElementTripletConvertTo("IN");
    element = document->FindNextElement("location");
  }

// Find all POINTMASS elements that descend from this METRICS branch of the
// config file.

  element = document->FindElement("pointmass");
  while (element) {
    AddPointMass(element);
    element = document->FindNextElement("pointmass");
  }

  double ChildFDMWeight = 0.0;
  for (int fdm=0; fdm<FDMExec->GetFDMCount(); fdm++) {
    if (FDMExec->GetChildFDM(fdm)->mated) ChildFDMWeight += FDMExec->GetChildFDM(fdm)->exec->GetMassBalance()->GetWeight();
  }

  Weight = EmptyWeight + in.TanksWeight + GetTotalPointMassWeight()
    + in.GasMass*slugtolb + ChildFDMWeight;

  Mass = lbtoslug*Weight;
```
For modeling mass properties on another planet, I'll need to change the lbstoslugs value so that specifying the weight of a craft on another planet converts to mass properly. Actually, it's probably more sensible to specify mass in the first place, in units of ONLY EITHER slugs or kilograms. 

I'm wondering, though, if this might break some peoples' models?

**Reply by seanmcleod** [reply]
> I'm wondering, though, if this might break some peoples' models?

If `planet == Earth` allow both `emptywt` and `emptymass`, but if not Earth then only allow `emptymass` and display an error if `emptymass` is specified? 

And go through all the aircraft in the JSBSim repo and convert them to `emptymass` to set an example going forward?

There are a couple of other elements though that will also need converting, e.g. fuel tank's `capacity` and `contents`.

**Reply by jonsberndt** [reply]
All true.

So, my thinking is that the user should be able to supply EITHER `<emptywt>` OR `<emptymass>`. Mass would be specified ONLY in slugs or kg, weight ONLY in lbs or Newtons. The weight given (if given) would be the weight on the planet chosen - not Earth weight.

This would probably deprecate current practice in some models, but I think it's the right way to go.

**Reply by seanmcleod** [reply]
Busy modeling the SpaceX Starship on earth and mars? :wink:

**Reply by bcoconni** [reply]
@seanmcleod 
> If `planet == Earth` allow both `emptywt` and `emptymass`, but if not Earth then only allow `emptymass` and display an error if emptymass is specified?

I guess your very last statement contains a typo and that you mean to display an error if `emptywt` (not `emptymass`) is specified outside Earth ?

@jonsberndt 
> So, my thinking is that the user should be able to supply EITHER `<emptywt>` OR `<emptymass>`. Mass would be specified ONLY in slugs or kg, weight ONLY in lbs or Newtons. The weight given (if given) would be the weight on the planet chosen - not Earth weight.

I completely agree with both of your statements. Honestly, I have always assumed that `<emptywt>` was improperly named and actually meant empty mass. I am working in the aerospace industry for 15 years and I have only seen specifications in mass (lbs or kg), but never ever in weight (lbf or N), even when I worked with UK or US companies. So my point of view is that your proposal would clarify the intent of that data but I'd really be very surprised that anyone would want to specify the weight of a spacecraft in the gravity of Mars (or any other planet by the way) :astonished: So yes it's a good idea to raise an error if `emptywt` is specified on a planet other than Earth.

**Reply by seanmcleod** [reply]
Yep typo in my last statement.

**Comment by jonsberndt** [comment]
Furthermore, all this has lead me to consider that it might be time for a FGPlanet class that models a user-selectable planet. It would include gravity, geodesy, and perhaps even atmosphere. If that is too large a change in code, it might be accomplished by specifying a planet to be used (Earth being default, of course) and then letting the code construct the correct planetary gravity, geodesy, and atmosphere.

**Reply by seanmcleod** [reply]
There is already some support for planet parameters, e.g.

https://github.com/JSBSim-Team/jsbsim/commit/5fe0882ac82bdfedb6f35c31f63a9660de82bf44

https://github.com/JSBSim-Team/jsbsim/commit/52516f5738c890f312f4ea43b74b9b3fc23354be

https://github.com/JSBSim-Team/jsbsim/commit/ecb79520a50d0f04ca095801f39c3bdb3ea36a16

**Reply by bcoconni** [reply]
The atmosphere of other planets still cannot be specified easily in JSBSim. I guess all we need would be to specify the "sea level" pressure and temperature as well as the temperature table (which is currently hard coded in `src/models/atmosphere/FGStandardAtmosphere.cpp`)
https://github.com/JSBSim-Team/jsbsim/blob/b3bc604500efbef1341c789d229b6009cd7d1c94/src/models/atmosphere/FGStandardAtmosphere.cpp#L84-L95
Also I think that `FGStandardAtmosphere` uses hard coded values for the gravity but we would need to use the planet gravity from `FGInertial` to allow the correct computation of the atmosphere pressure and temperature along the altitude.
